### PR TITLE
Patmat: Fix over-pruning of private children

### LINKED ISSUE #12712: Incorrect "unreachable code" warning
Using 2.13.10

```scala
object T {
  private sealed trait T
  private object O extends T
  private trait U extends T
  private object P extends U

  def t(t: T) = t match {
    case O => ()
    case _: U => println("hai")
  }

  def main(args: Array[String]): Unit = {
    t(P)
  }
}
```

gives

```
➜ sandbox sc Test.scala
Test.scala:9: warning: unreachable code
    case _: U => println("hai")
                        ^
1 warning

➜ sandbox s T
hai
```

The warning goes away when `U` is changed to `sealed`

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes https://github.com/scala/bug/issues/12712

### CODE REVIEW

**Review by lrytz — CHANGES_REQUESTED** [review]
When I change the example a little I now get a false exhaustivity warning:

```scala
object T {
  private sealed trait T
  private object O extends T
  private trait U extends T
  private object P extends U

  def t1(t: T) = t match {
    case O => ()
    case _: U => println("hai") // false "unreachable code" in 2.13.10
  }

  def t2(t: T) = t match {
    case O => ()
    case P => println("hai") // false "would fail on (_: T.U)" with this PR
  }
}
```

**Comment by dwijnand** [comment]
That's not false:

```scala
object Outlaw extends U
t2(Outlaw)
```

`T` is the end of the line, because it isn't sealed.  Unless you want to extend the sealed child machinery so that "private" is more "effectively sealed" and has sealed children.  But I don't think we should.

**Comment by lrytz** [comment]
Ah, so `private` is not taken as effectively sealed. I haven't thought about this much, why do you think that's the right thing to do?

**Comment by dwijnand** [comment]
Well the machinery for children is currently only set up - I mean in the compiler in general - for sealed classes.  So we'd have to extend that.  And I'm not sure if extending that is an expected/intended/acceptable change?  Maybe just yes, maybe sealed should be reserved for that.

**Comment by som-snytt** [comment]
I agree about "effectively sealed", because people (we the users) often reason from the obvious, such as, "it knows I don't have a subclass." I remember something about sealed and children from looking at java sealed, but to me, sealed is a stronger mechanism than warnings due to "effectively sealed." However, I don't have a lot of experience (history of annoyance) to fall back on.

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala [hunk 1]
```diff
@@ -129,8 +129,17 @@ trait TreeAndTypeAnalysis extends Debugging {
       def filterAndSortChildren(children: Set[Symbol]) = {
         // symbols which are both sealed and abstract need not be covered themselves,
         // because all of their children must be and they cannot otherwise be created.
-        children.toList.filter(x => !(x.isSealed || x.isPrivate) || !x.isAbstractClass || isPrimitiveValueClass(x))
-          .sortBy(_.sealedSortName)
+        val children1 = children.toList.filter { child =>
+          !child.isSealed ||           // keep non-sealed children
+          !child.isAbstractClass ||    // keep sealed, non-abstract children
+          isPrimitiveValueClass(child) // keep sealed, abstract, primitive classes
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: This is redundant I think, value classes are not sealed (they are abstract though).

**Review by lrytz — COMMENTED** [review]
Finally got it... removing `isPrivate` is the fix for the bug, but that was an incorrect fix for t6159. So you added a fix for t6159 below.

I changed things a bit, maybe it's clearer?

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala [hunk 2]
```diff
@@ -129,16 +129,10 @@ trait TreeAndTypeAnalysis extends Debugging {
...
-        }.sortBy(_.sealedSortName)
-        children1.filter { child =>
-          !child.isAbstractClass || // keep non-abstract children
-          !children1.exists(sym => (sym ne child) && sym.isSubClass(child))
-          // keep abstract children that aren't superclasses of other children
-          // for example in t6159 drop X2
+        val children1 = children.toList.filterNot(child => child.isSealed && child.isAbstractClass).sortBy(_.sealedSortName)
+        children1.filterNot { child =>
+          // remove abstract children that are superclasses of other children, for example in t6159 drop X2
+          child.isAbstractClass && children1.exists(sym => (sym ne child) && sym.isSubClass(child))
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: maybe we can only remove private children here? This example (two files) no longer shows an exhaustivity warning:

```
object A {
  sealed trait X
  class X1 extends X2 with X { }
  abstract class X2 extends X
  sealed trait X3 extends X

  def f(x: X) = x match {
    case _: X1 => 0
  }
}
```

```
object B {
  def main(args: Array[String]): Unit = {
    A.f(new A.X2 { })
  }
}

```

File: test/files/pos/t6159.scala [hunk 3]
```diff
@@ -9,3 +9,16 @@ trait A {
...
+// like A2, but with T2 not private
+trait A2 {
+  sealed abstract class X
+  private class X1 extends X with X2 { }
+  trait X2 extends X
+  sealed trait X3 extends X
+
+  def f(x: X) = x match {
+    case _: X1 => 0
+    case _: X2 => 1 // required, otherwise inexhaustive
```
Comment (lrytz) [inline test/files/pos/t6159.scala]: maybe put this in a `neg/` test..

**Comment by SethTisue** [comment]
> remove private abstract children

I was something of a private abstract child, myself

**Review by lrytz — APPROVED** [review]

**Pull request merged**